### PR TITLE
Adaptive dynamic number of speculative tokens

### PR DESCRIPTION
**Pull request by jmamou**
# What does this PR do?
Following
https://github.com/huggingface/transformers/pull/33258
https://github.com/huggingface/transformers/pull/33657

The assistant's confidence threshold is adjusted throughout the speculative iterations to reduce the number of unnecessary draft and target forward passes. The costs are estimated based on the ROC curve, which considers the probability of the draft token and its match with the target. A cost of 25% is assigned to false positives and 75% to false negatives.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@gante @amyeroberts

### CODE REVIEW

File: src/transformers/generation/candidate_generator.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 import numpy as np
 import torch
+from sklearn.metrics import roc_curve
```
Comment (gante) [inline src/transformers/generation/candidate_generator.py:21]: sklearn is not installed by default, and we don't want to introduce a hard dependency to use assisted generation.

1 - we'll have to wrap this import in an import guard, i.e.
```py
from ..utils import is_sklearn_available
if is_sklearn_available():
    from sklearn.metrics import roc_curve
```
otherwise folks without the library will fail to initialize this module
2 - given that's an optional import and that this PR's technique "only" boosts speed, my suggestion would be to use the boost if the library is available, and don't use it if not. That way, assisted generation always works and the user is not forced to install an extra dependency 🤗 In the docs, we should recommend to install sklearn to benefit from this boost.

**Review by gante — COMMENTED** [review]
In addition to the comments: can you share the benchmark results here as well, for future reference?

**Comment by gante** [comment]
@jmamou yeah, let's please run more benchmarks before merging. 

In the odd chance it ends up being beneficial only in very specific circumstances, I'd rather **not** merge the technique to avoid adding complexity (which usually reduces our team's ability to work on more projects 🤗 )

**Comment by jmamou** [comment]
> @jmamou yeah, let's please run more benchmarks before (iterating on the PR and) merging
> 
> In the odd chance it ends up being beneficial only in very specific circumstances, I'd rather **not** merge the technique to avoid adding complexity (which usually reduces our team's ability to work on more projects 🤗 )

@gante  
I have run benchmarks from  https://huggingface.co/spaces/joaogante/assisted_generation_benchmarks
https://github.com/gante/huggingface-demos/tree/main/experiments/faster_generation
Evaluated metric: throughput -- time per token in ms, lower is better
Device: RTX 3090; dtype applies to both models


Model | Assistant | dtype | task | sampling? | w/o assistant | disco 0.4 | adaptive disco | disco speedup | adaptive disco speedup
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
openai/whisper-large-v2 | openai/whisper-tiny | fp16 | automatic speech recognition | no | 20.02 | 14.59 | 13.81 | 1.37 | 1.45
facebook/opt-6.7b | facebook/opt-125m | bf16 | summarization | no | 23.81 | 8.73 | 8.72 | 2.73 | 2.73
facebook/opt-6.7b | facebook/opt-125m | bf16 | summarization | yes (t=0,6) | 24.21 | 12.01 | 10.55 | 2.02 | 2.29
facebook/opt-6.7b | facebook/opt-125m | bf16 | open-ended generation | no | 22.14 | 14.19 | 14.14 | 1.56 | 1.57
facebook/opt-6.7b | facebook/opt-125m | bf16 | open-ended generation | yes (t=0,7) | 22.13 | 14.16 | 14.09 | 1.56 | 1.57
Salesforce/codegen-6B-mono | Salesforce/codegen-350M-mono | bf16 | code generation (python) | no | 30.88 | 26.8 | 26.95 | 1.15 | 1.15
Salesforce/codegen-6B-mono | Salesforce/codegen-350M-mono | bf16 | code generation (python) | yes (t=0,4) | 37.02 | 35.88 | 33.79 | 1.03 | 1.1
google/flan-t5-xl | google/flan-t5-small | bf16 | summarization | no | 24.76 | 20.11 | 20.1 | 1.23 | 1.23
google/flan-t5-xl | google/flan-t5-small | bf16 | summarization | yes (t=0,6) | 24.44 | 26.78 | 25.15 | 0.91 | 0.97

Model | Assistant | dtype | task | sampling? |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
meta-llama/Llama-3.1-8B | meta-llama/Llama-3.2-1B | bf16 | summarization | no | 33.06 | 19.27 | 19.29 | 1.72 | 1.71
meta-llama/Llama-3.1-8B | meta-llama/Llama-3.2-1B | bf16 | summarization | yes (t=0,6) | 33.6 | 24.35 | 21.69 | 1.38 | 1.55
meta-llama/Llama-3.1-8B | meta-llama/Llama-3.2-1B | bf16 | open-ended generation | no | 31.25 | 33.2 | 33.1 | 0.94 | 0.94
meta-llama/Llama-3.1-8B | meta-llama/Llama-3.2-1B | bf16 | open-ended generation | yes (t=0,7) | 31.35 | 42.29 | 39.02 | 0.74 | 0.8
meta-llama/Llama-3.1-8B | meta-llama/Llama-3.2-1B | bf16 | code generation (python) | no | 27.98 | 19.49 | 19.72 | 1.44 | 1.42
meta-llama/Llama-3.1-8B | meta-llama/Llama-3.2-1B | bf16 | code generation (python) | yes (t=0,4) | 28.6 | 24.23 | 20.85 | 1.18 | 1.37


An improvement is observed when `do_sample=True`, likely because the threshold was set to 0.4 to optimize for greedy decoding. It seems that a lower threshold may be needed when sampling, highlighting the need to adapt the threshold as proposed in the PR ...

**Comment by gante** [comment]
@jmamou I'm convinced :D The benchmarks do show a consistent upgrade

File: src/transformers/generation/candidate_generator.py [hunk 2]
```diff
@@ -180,6 +186,10 @@ def __init__(
         # We need to roll back the cache in assisted generation, only DynamicCache is supported
         self.generation_config.cache_implementation = None
 
+        if self.assistant_model.generation_config.assistant_confidence_threshold:
```
Comment (gante) [inline src/transformers/generation/candidate_generator.py:189]: This block can be conditional to `is_sklearn_available`, correct? (if we don't have sklearn we don't adapt the threshold, and therefore we don't need to store this data)

File: src/transformers/generation/candidate_generator.py [hunk 3]
```diff
@@ -230,6 +240,12 @@ def get_candidates(self, input_ids: torch.LongTensor) -> Tuple[torch.LongTensor,
         # 3. Update variables for the next round of candidate generation
         self.assistant_kwargs["past_key_values"] = assistant_output.past_key_values
 
+        if self.assistant_model.generation_config.assistant_confidence_threshold:
```
Comment (gante) [inline src/transformers/generation/candidate_generator.py:243]: Same comment as above

**Review by gante — APPROVED** [review]
LGTM, thank you for the thorough benchmark 🤗

File: src/transformers/generation/candidate_generator.py (hunk 1 above)
Comment (jmamou) [reply src/transformers/generation/candidate_generator.py]: @gante where do you suggest to  add the recommendation in the docs? 
[here](https://huggingface.co/docs/transformers/main/en/generation_strategies#speculative-decoding) ?

File: src/transformers/generation/candidate_generator.py [hunk 4]
```diff
@@ -230,6 +240,12 @@ def get_candidates(self, input_ids: torch.LongTensor) -> Tuple[torch.LongTensor,
         # 3. Update variables for the next round of candidate generation
         self.assistant_kwargs["past_key_values"] = assistant_output.past_key_values
 
+        if is_sklearn_available() and self.assistant_model.generation_config.assistant_confidence_threshold:
+            for i, score in enumerate(assistant_output.scores):
+                id = assistant_output.sequences[-1, i - len(assistant_output.scores)]
+                p = score.softmax(-1)[0, id].item()
+                self.probs.append(p)
```
Comment (keyboardAnt) [inline src/transformers/generation/candidate_generator.py:247]: Consider vectorizing

**Comment by jmamou** [comment]
@ArthurZucker could you please review it?

File: src/transformers/generation/candidate_generator.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/generation/candidate_generator.py]: @jmamou yes, that will be the correct place in docs

**Review by ArthurZucker — COMMENTED** [review]
LGTM! It's just missing a test !

**Review by ArthurZucker — COMMENTED** [review]
LGTM! It's just missing a test !

**Comment by jmamou** [comment]
> LGTM! It's just missing a test !

@ArthurZucker 
done!

**Comment by ArthurZucker** [comment]
`from transformers.generation.candidate_generator import AssistedCandidateGenerator` needs to protect it's import to torch! `candidate_generator.py` needs to check python availability!

**Comment by ArthurZucker** [comment]
Let's GOOOOOOO! 🚀
